[sw/dif] Implement a DIF library for HMAC

jon-flatley opened the pull request:
This is my first attempt to write a DIF (or anything for that matter)  for the opentitan project. This PR has @mcy's SPI DIF #1135 as a dependency since it adds some volatile register access abstractions that I wanted to use. 

There are a few bits that I'm unhappy with, like how groups of registers at adjacent addresses are handled, but I was trying to stay in the spirit of @gkelly's DIF document and not make any assumptions about register addresses. I'm more than happy to change these things if anyone has better ideas to offer.

I also included some ragtag smoke test cases that I used to convince myself that this DIF actually does useful things. This will probably get replaced with some more useful testing once we have a DIF testing process worked out.

alphan commented:
Hi @jon-flatley, since #1314 has landed, you can rebase, use those functions and remove other files.

jon-flatley commented:
Thanks for the tip @alphan, refactored version is now pushed and the dependency on @mcy's SPI DIF has been removed.

silvestrst commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,138 @@
+ */
+typedef enum dif_hmac_ops {
+  DIF_HMAC_OP_HMAC = 0,
+  DIF_HMAC_OP_SHA256 = 1
```
The format for the constants is `kDifHmacOpSha256`, consider naming it `kDifHmacOpsSha256` instead to match the type name.

I don't really like camel case in C, but all the rest of the OT uses this stye, so we should stick with it. And that is a preferred way described in the style guide.

silvestrst commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,138 @@
+/**
+ * A typed representation of the HMAC digest.
+ */
+typedef struct dif_hmac_digest { uint32_t digest[8]; } dif_hmac_digest_t;
```
Could you run `git clang-format` (I think only considers `git add` added files) or `clang-format` on all changed files.

jon-flatley replied:
This formatting was actually produced by clang-format on the Azure pipeline. My local clang-format doesn't produce this, but I was taking the Azure pipeline checks as the source of truth.

```
diff --git a/sw/device/lib/dif/dif_hmac.h b/sw/device/lib/dif/dif_hmac.h
index 25175a8e..7288c5f1 100644
--- a/sw/device/lib/dif/dif_hmac.h
+++ b/sw/device/lib/dif/dif_hmac.h
@@ -37,9 +37,7 @@ typedef struct dif_hmac_config {
 /**
  * A typed representation of the HMAC digest.
  */
-typedef struct dif_hmac_digest {
-  uint32_t digest[8];
-} dif_hmac_digest_t;
+typedef struct dif_hmac_digest { uint32_t digest[8]; } dif_hmac_digest_t;
 
 /**
  * State for a particular HMAC device.
@@ -47,9 +45,7 @@ typedef struct dif_hmac_digest {
  * its member variables should be considered private, and are only provided so
  * that callers can allocate it.
  */
-typedef struct dif_hmac {
-  mmio_region_t base_addr;
-} dif_hmac_t;
+typedef struct dif_hmac { mmio_region_t base_addr; } dif_hmac_t;
 
 /**
  * Initializes the SPI device described by |config|, writing internal state to
```

mcy replied:
I think this is a problem where the clang-format on our workstations might have slightly different defaults. I suspect Azure is just using an older clang-format. *shrug*

jon-flatley replied:
Seems that way. Azure and my `clang-format` disagree in this specific case, and I've been manually fixing this to appease Azure.

silvestrst commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,138 @@
+ * used to "randomize" the internal state of the HMAC device. See the HMAC IP
+ * documentation for more information: hw/ip/hmac/doc.
+ *
+ * @param hmac the HMAC device to clobber state on.
```
`@param hmac the HMAC device to clobber state on.` could you start the variable description with a capital, like `@param hmac The ...`.

silvestrst commented on sw/device/tests/hmac/dif_hmac_test.c:
```diff
@@ -0,0 +1,116 @@
+  uart_send_str(" bits.\r\n");
+
+  while (!dif_hmac_is_done(&hmac))
+    ;
```
Is this something that `clang-format` done, or just a typo?

mcy replied:
I think clang-format did this. Either way, it should be `{}`, not a semi.

mcy commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,138 @@
+  /** Message endianness. true -> big endian, false -> little endian. */
+  bool endian_swap;
+  /** Digest endianness. true -> big endian, false -> little endian. */
+  bool digest_swap;
```
Consider adding an endianness enum.

mcy commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,138 @@
+/**
+ * Supported HMAC operations.
+ */
+typedef enum dif_hmac_ops {
```
Are you sure you don't want this to be named `dif_hmac_mode`?

mcy commented on sw/device/lib/dif/meson.build:
```diff
@@ -16,3 +16,16 @@ dif_uart = declare_dependency(
 )
+
+# HMAC device DIF library
+sw_dif_hmac= declare_dependency(
```
Missed a space.

mcy commented on sw/device/tests/hmac/dif_hmac_test.c:
```diff
@@ -0,0 +1,116 @@
+    data += dif_hmac_push_fifo(&hmac, data, kDataLen - (data - kData));
+  }
+  for (volatile uint32_t i = 0; i < 10000; i++)
+    ;
```
If you want to burn cycles, use one of the sleep functions in lib/runtime.

jon-flatley replied:
Oh dear, not only did clang-tidy do some funky stuff here but I evidently left in some debugging code. I'll clean this up.

mcy commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,140 @@
+  uint64_t msg_upper =
+      mmio_region_read32(hmac->base_addr, HMAC_MSG_LENGTH_UPPER(0));
+
+  return (msg_upper << (sizeof(uint32_t) * 8)) | msg_lower;
```
I wonder if we want to add this type of operation to a header in libbase (e.g., u64_from_u32s())...

jon-flatley commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,140 @@
+  if (config->mode == DIF_HMAC_OP_HMAC) {
+    device_config |= 1 << HMAC_CFG_HMAC_EN;
+
+    // Set the key making no assumptions about register locations.
```
Since the `HMAC_KEY#_REG_OFFSET` values are declared individually and reflect the offsets in the documentation I was taking the safest approach by not making any assumptions about the register layout. If this isn't a realistic concern I'll happily make this a loop.

jon-flatley commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,140 @@
+  while ((words_remaining || bytes_remaining) && fifo_avail(hmac) > 0) {
+    if (words_remaining > 0) {
+      // Write full words to the FIFO
+      mmio_region_write32(hmac->base_addr, HMAC_MSG_FIFO(0), *wp++);
```
I think you're right, this violates aliasing rules. I'm refactoring this loop per your other comments and will fix this.

jon-flatley commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,140 @@
+  uint8_t bytes_remaining = len % sizeof(uint32_t);
+
+  // Write data until we either finish or the FIFO fills up.
+  while ((words_remaining || bytes_remaining) && fifo_avail(hmac) > 0) {
```
Fair enough, I'll split this up into separate loops

jon-flatley commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,140 @@
+}
+
+size_t dif_hmac_push_fifo(const dif_hmac_t *hmac, void *data, size_t len) {
+  const uint32_t *wp = (uint32_t *)data;
```
Is the memory.h `memcpy` preferable to `__builtin_memcpy` in this situation?

jon-flatley replied:
I see the answer to this question in the DIF Implementations Questions doc, I'll use memory.h

lenary replied:
Yes, please do use libbase's `memory.h`. It makes the dependencies clearer.

jon-flatley commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,140 @@
+      words_remaining--;
+    } else {
+      // Write the remaining bytes to the FIFO.
+      uint8_t *bp = (uint8_t *)wp;
```
I'm surprised this doesn't generate a compiler warning. As a side effect of the refactoring necessary based on your other comments this is fixed

gkelly commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,208 @@
+  while (words_remaining > 0 && get_fifo_available_space(hmac) > 0) {
+    uint32_t word;
+
+    memcpy(&word, data_sent, sizeof(uint32_t));
```
This has to be `__builtin_memcpy`, `memcpy` isn't available to DIFs.

jon-flatley replied:
I originally was using `__builtin_memcpy` but [this conversation](https://github.com/lowRISC/opentitan/pull/1320#discussion_r368614772) changed my mind.

gkelly commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,208 @@
+  }
+
+  return kDifHmacNoError;
+}
```
The semantics here are that this function will (except for invalid arguments) always return `kDifHmacNoError`, even if no bytes were sent.  I'm thinking about if those are the correct semantics.  I could envision a caller doing:

```
char *buffer = ...;
size_t buffer_size = N;

size_t written = 0;
while (written < buffer_size) {
  size_t bytes_sent;
  const dif_hmac_error_t err = dif_hmac_fifo_push(hmac, buffer + written, buffer_size - written, &bytes_sent);
  if (err != kDifHmacNoError) {
    // Error reporting.
    break;
  }
}
```
Maybe I'm being paranoid here.  If the HMAC is running then it shouldn't stay full forever.  I guess if a caller tries to add data and the HMAC isn't running then this could block forever.

I think this function should return an error when it can write no data.  My justification: The function is `dif_hmac_fifo_push`, and if it pushed no data then it did not do the thing it's named.

jon-flatley replied:
I like this, especially with the argument that these functions should be getting their own error types. In practice the copy process seems slow enough that the FIFO never actually fills up under this implementation, so it makes even more sense to return an error if the buffer isn't consumed in full.

gkelly commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,208 @@
+  return kDifHmacNoError;
+}
+
+dif_hmac_error_t dif_hmac_get_msg_len(const dif_hmac_t *hmac,
```
`dif_hmac_message_length`

gkelly commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,208 @@
+  return kDifHmacNoError;
+}
+
+dif_hmac_error_t dif_hmac_is_done(const dif_hmac_t *hmac, bool *done) {
```
`dif_hmac_done`

gkelly commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,208 @@
+  return kDifHmacNoError;
+}
+
+dif_hmac_error_t dif_hmac_read_digest(const dif_hmac_t *hmac,
```
`dif_hmac_digest_read`

gkelly commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,162 @@
+ * Initializes the SPI device described by |config|, writing internal state to
+ * |hmac_out|.
+ *
+ * This function *must* be called on a particular |reg32_t| before calling any
```
Fixup references to `reg32_t`.

gkelly commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,162 @@
+typedef struct dif_hmac { mmio_region_t base_addr; } dif_hmac_t;
+
+/**
+ * Initializes the SPI device described by |config|, writing internal state to
```
HMAC

gkelly commented on sw/device/tests/hmac/dif_hmac_test.c:
```diff
@@ -0,0 +1,150 @@
+  // Some out-params
+  size_t sent_bytes;
+  uint64_t sent_bits;
+  bool hmac_running = true;
```
Move defs as close as possible to their first use site.

jon-flatley commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,208 @@
+ * A helper funciton for calculating |HMAC_FIFO_MAX| - |get_fifo_entry_count()|.
+ */
+static uint32_t get_fifo_available_space(const dif_hmac_t *hmac) {
+  return 16;
```
Left over debugging, yep. Thanks for pointing this out.

jon-flatley commented:
Pushed a new revision. A few changes:

- Cleaned up the tests considerably. Originally they were just for my own sanity, now they're a bit more crisp.
- Added `dif_hmac_fifo_count_entries()` as this was important missing functionality.
- Added more descriptive error types.
- Addressed other typos and refactoring requests made in this thread.

jon-flatley commented:
And another small push to fix a bug in the test #1449.

alphan commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,221 @@
+}
+
+dif_hmac_error_t dif_hmac_wipe_secret(const dif_hmac_t *hmac,
+                                      uint32_t entropy) {
```
I wonder if the function should enforce `entropy != 0`. Does it matter? If so, we should consider making `entropy` an optional parameter and using a hw counter or something else if it is missing or zero.

jon-flatley replied:
I don't think that kind of enforcement is necessary. 
> The software can wipe the variables by writing a 32-bit random value into WIPE_SECRET register. The internal variables will be reset to the written value. This version of the HMAC doesn’t have a internal pseudo-random number generator to derive the random number from the written seed number.
So I think writing 0 may be valid. Also 0 is a valid output from most entropy sources.

I'm not sure it's practical to use architecture features such as hw counters since DIFs are meant to be architecture independent and dependency free.

alphan replied:
> since DIFs are meant to be architecture independent and dependency free.

Good point.

alphan commented on sw/device/tests/hmac/dif_hmac_test.c:
```diff
@@ -0,0 +1,190 @@
+    uart_send_str("PASS!\r\n");
+  } else {
+    uart_send_str("FAIL!\r\n");
+  }
```
A minor point but it would be cleaner if we had a wrapper around `run_test` that takes also a test name and handles success and failure internally. You can consider renaming `run_test` to `run_test_impl` and write `run_test()` like
```
void run_test(const char *test_name, const dif_hmac_config_t *config, const char *data,
             size_t len, const uint32_t *expected_digest) {
  uart_send_str("\r\n");
  uart_send_str(test_name);
  uart_send_str("\r\n");
  if (run_test(config, data, len, expected_digest) {
    uart_send_str("PASS!\r\n");
  } else {
    uart_send_str("FAIL!\r\n");
  }
}

alphan commented on sw/device/tests/hmac/dif_hmac_test.c:
```diff
@@ -0,0 +1,190 @@
+  hmac_config.base_addr = mmio_region_from_addr(HMAC0_BASE_ADDR);
+  hmac_config.digest_endianness = kDifHmacBigEndian;
+  hmac_config.message_endianness = kDifHmacBigEndian;
+  hmac_config.mode = kDifHmacModeSha256;
```
How about inline initialization?
```
dif_hmac_config_t hmac_config {
  .base_addr = mmio_region_from_addr(HMAC0_BASE_ADDR),
  .digest_endianness = kDifHmacBigEndian,
  .message_endianness = kDifHmacBigEndian,
  .mode = kDifHmacModeSha256
};
```

jon-flatley commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,211 @@
+ */
+typedef enum dif_hmac_mode {
+  /** The HMAC mode */
+  kDifHmacModeHmac = 0,
```
Matter of habit mostly. For the error enums it could simplify certain types of error checking, but conversations about leveraging the falsiness of `kOk = 0` for error checking seem to end in the consensus that it's a bad idea. I don't think there is any real reason to keep these enum-constants.

jon-flatley commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,211 @@
+ */
+typedef enum dif_hmac_endianness {
+  /** Big endian byte ordering */
+  kDifHmacBigEndian = 0,
```
Is there some style guide or some other precedent you can cite that suggests prefixing enum fields with the full name of the enum itself? [What I think is the relevant style guide](https://google.github.io/styleguide/cppguide.html#Enumerator_Names) makes no apparent mention of this.

alphan replied:
Yes, there is a couple. For instance, please see: @gkelly's comment [here](https://github.com/lowRISC/opentitan/pull/1254#discussion_r374303741)

silvestrst replied:
@alphan +1

jon-flatley replied:
See my response to @gkelly's comment [here](https://github.com/lowRISC/opentitan/pull/1254#discussion_r375904676). I think this needs to go in the DIF style guide.

alphan replied:
#1527

jon-flatley commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,221 @@
+}
+
+/**
+ * Returns the number of entries in the FIFO of |hmac|. If the FIFO is empty,
```
It depends on what's written. In the case where a 32-bit word is written, an entry is 32 bits. In the case that a single byte is written, the entry is 8 bits. At least this is how I understand the specification and it seems to produce the correct hash.

alphan replied:
Thanks for the explanation! Actually, I read the spec before reviewing this DIF but somehow missed this. The following part from the spec sums it up nicely, I think:

> The message length is calculated by the packer logic. The packer converts non-word writes into full word writes and feeds into the message FIFO.

jon-flatley commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,221 @@
+  uint64_t msg_upper =
+      mmio_region_read32(hmac->base_addr, HMAC_MSG_LENGTH_UPPER_REG_OFFSET);
+
+  *msg_len = (msg_upper << (sizeof(uint32_t) * 8)) | msg_lower;
```
Would replacing `8` with `CHAR_BIT` be an acceptable replacement? I prefer keeping `sizeof()` instead of just a constant since it's more clear where the value comes from.

alphan replied:
I take my comment back since this is mostly a matter of style and we may eventually introduce a utility function for this operation.

lenary commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,208 @@
+  return HMAC_FIFO_MAX - get_fifo_entry_count(hmac);
+}
+
+dif_hmac_error_t dif_hmac_init(const dif_hmac_config_t *config,
```
The problem with per-function error enums is:
- They are hard to stringify (you end up with many error code to string functions)
- You get lots of repeated enum variants for `Success`, `BadArg` (or equivalent), or the kind of general error that hitting an enum switch's `default` should reach (where the `default` is being used for representable but illegal enum values).

But I agree they're much better for refactoring.

mcy replied:
Both of my DIF libraries take Garret's suggested approach. It looks horrific, frankly, but I'm not sure there's a better solution.

jon-flatley commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,221 @@
+  // Set the byte-order of the input message.
+  uint32_t device_config = 0;
+  if (config->message_endianness == kDifHmacBigEndian) {
+    device_config |= (1 << HMAC_CFG_ENDIAN_SWAP);
```
I did consider it, but I chose to write the full config to the register. I don't really see much of a difference. I'd argue that the name `device_config` adds a bit more context, and as you mentioned it reduces the number of mimo accesses.

jon-flatley commented:
Thanks @alphan for the comments. I've resolved a few and responded to those that I didn't address in my latest push.

mcy commented:
Now that it's possible to write tests for DIFs, I get to say "please include tests with your change." =D

Silvestrs wrote a pretty good first-shot you can use as a template: #1477

alphan commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,221 @@
+#include "hmac_regs.h"  // Generated.
+#include "sw/device/lib/base/memory.h"
+
+#define HMAC_FIFO_MAX (HMAC_MSG_FIFO_SIZE_BYTES / 32)
```
Assuming that you want the size in words here, shouldn't this be `HMAC_MSG_FIFO_SIZE_BYTES / 4` since `HMAC_MSG_FIFO_SIZE_BYTES` is in bytes? Actually, `hmac_regs.h` already defines `HMAC_MSG_FIFO_SIZE_WORDS` so it may be a better idea to use it directly.

jon-flatley replied:
What a mess, yes I should be using `HMAC_MSG_FIFO_SIZE_WORDS`, thank you.

alphan commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,221 @@
+      config->digest_endianness > kDifHmacLittleEndian ||
+      config->message_endianness < kDifHmacBigEndian ||
+      config->message_endianness > kDifHmacLittleEndian ||
+      config->mode < kDifHmacModeHmac || config->mode > kDifHmacModeSha256) {
```
Thanks @silvestrst (for some reason I cannot see your comment on github so replying to my own). The edit was about the correctness of the checks. These checks are logically correct but they still should be removed (see the quote from DIF style guide above).

mcy replied:
I think some diffs got dropped on the ground here?

silvestrst commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,221 @@
+  if (words_remaining > 0 || bytes_remaining > 0) {
+    return kDifHmacFifoFull;
+  }
+  return kDifHmacFifoNoError;
```
Nit: could you add a newline before this line.

silvestrst commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,221 @@
+  return kDifHmacNoError;
+}
+
+dif_hmac_fifo_error_t dif_hmac_fifo_push(const dif_hmac_t *hmac,
```
This routine looks slick, thank you for refactoring.

silvestrst commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,221 @@
+    return kDifHmacInvalidArgument;
+  }
+
+  *fifo_depth = get_fifo_entry_count(hmac);
```
Nit: could you add a newline after this line?

silvestrst commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,211 @@
+ */
+typedef enum dif_hmac_error {
+  /** No error occurred. */
+  kDifHmacNoError = 0,
```
or maybe `kDifHmacErrorNone` maybe has a little better ring to it? What do you guys think?

jon-flatley replied:
@silvestrst I prefer `kDifHmacErrorNone`

silvestrst commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,211 @@
+ */
+typedef enum dif_hmac_fifo_error {
+  /** No error occurred. */
+  kDifHmacFifoNoError = 0,
```
or maybe `kDifHmacFifoErrorNone` maybe has a little better ring to it? What do you guys think?

silvestrst commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,211 @@
+ * @param hmac The HMAC device to send to.
+ * @param data A contiguous buffer to copy from.
+ * @param len The length of the buffer to copy from.
+ * @param The number of bytes sent to the FIFO (optional).
```
`@param bytes_sent ...`

This discussion is outside of the scope of this review, but:
I really think it would be nice to have a tool that just verifies that function signatures match the function description, I suggested that to @asb and @lenary , just something to think about.

silvestrst commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,211 @@
+ * message length.
+ *
+ * @param hmac The HMAC device to get the FIFO depth for.
+ * @param fifo_len The number of entries in the FIFO.
```
`@param fifo_depth ...`

mcy commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,221 @@
+  return HMAC_FIFO_MAX - get_fifo_entry_count(hmac);
+}
+
+dif_hmac_error_t dif_hmac_init(const dif_hmac_config_t *config,
```
I think a more realistic scenario is having Tock multiplex HMAC requests from processes using different keys. I would expect plain SHA-256 to be a separate DIF?

jon-flatley replied:
> Is it possible that HMAC could be reconfigured during runtime? As it has two modes, "simple" HASH and "secure" hash, a scenario could potentially be: most of the time use it just as a checksum, sometimes use it as both.
> 
> In that case I think a separate `configure` routine might be beneficial. Could be done similar to UART DIF, where there is a static `configure` function, and two public API `dif_..._init` and `dif_..._configure`? What do you think?

Absolutely, but my thinking is that the SHA256 and HMAC use cases are different enough that it warrants initializing a new `dif_hmac_t`. After all, `dif_hmac_init` takes a `dif_hmac_config_t` so in that sense it _is_ the `configure` routine. If there's a case for finer grain configuration then perhaps they should be added individually, like `dif_hmac_mode_set` or `dif_hmac_key_set`.


> I think a more realistic scenario is having Tock multiplex HMAC requests from processes using different keys. I would expect plain SHA-256 to be a separate DIF?

This would break with the pattern of having a DIF per IP block, and considering there's only a single bit difference between the two modes this seems like overkill. Are there other DIFs that support multiplexing?

alphan replied:
I think all of these requirements can be handled in the driver level.

lenary replied:
I had a chat with @gkelly and, upon reflection, we're not sure having more than one `dif_hmac_t` for the same MMIO region is particularly safe (it opens up too many concurrency issues).

This almost certainly suggests, as @silvestrst says, that there should be a way, in this DIF, to:
- choose a different HMAC mode, endianness configuration,
- enable disable both the HMAC datapath and the SHA datapath
- load in a specific secret key

These should be different from initialisation, and preferably different from each other.

mcy commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,221 @@
+    return kDifHmacFifoInvalidArgument;
+  }
+
+  const uint8_t *data_sent = (uint8_t *)data;
```
Put a `const` in the cast for good measure.

mcy commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,221 @@
+    memcpy(&word, data_sent, sizeof(uint32_t));
+    mmio_region_write32(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, word);
+    data_sent += sizeof(uint32_t);
+    words_remaining--;
```
Flip the -- around; style guide requires `--x` everywhere.

mcy commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,221 @@
+  int bytes_remaining = len % sizeof(uint32_t);
+
+  // Write whole words until the FIFO fills or there are no words left to send.
+  while (words_remaining > 0 && get_fifo_available_space(hmac) > 0) {
```
Why not add a third loop to bring `data_sent` into `uint32_t` alignment, and then use `read_32` from memory.h for the central loop?

jon-flatley replied:
Are you suggesting writing individual bytes until `data_sent` is aligned?

jon-flatley replied:
I don't think so. Let's say `len` is 7 bytes, and aligned as such
```
XXXXXX01 23456789 ABCDXXXX
      ^
  data_sent
```
Since `data_sent` ins't necessarily aligned to begin with I think it would require three loops to

1. Bring `data_sent` into alignment by writing bytes from the first, unaligned word `{ 0x01 }`
```
XXXXXX01 23456789 ABCDXXXX
         ^
     data_sent
```
2. Loop on aligned words with `read_32` to send all 32-bit words `{ 0x23456789 }`
```
XXXXXX01 23456789 ABCDXXXX
                  ^
              data_sent
```
3. Write the final bytes from the last word `{ 0xAB, 0xCD }`

@mcy is this what you're suggesting?

alphan replied:
(Moved my reply over email here)

Yes, you’re right. For some reason I only thought about the beginning not the end. Sorry for the confusion!

What do you think about the following snippet? Please feel free to ignore this, change constants to `sizeof`s, etc.. I know this goes against @silvestrst's earlier comment but somehow looks simpler to me.

```
const uint8_t *data_sent = (const uint8_t *)data;
int rem_bytes = len;

while(rem_bytes > 0 && get_fifo_available_space(hmac) > 0) {
  if ((data_sent & 0x3u) == 0 && rem_bytes >= 4) {
    // Word aligned
    uint32_t word = read_32(data_sent);
    mmio_region_write32(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, word);
    data_sent += 4;
    rem_bytes -= 4;
  } else {
    mmio_region_write8(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, *data_sent);
    ++data_sent;
    --rem_bytes;
  }
}
```

jon-flatley replied:
I do like the idea of having a single loop, and I like how it reduces code duplication by combining the word alignment and trailing byte writes into a single case. I think the way it is now, with the three independent loops, is easier to reason about since the three steps always happen in the same order and the code makes that clear.

mcy commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,211 @@
+  /** Byte endianness for reads from the digest */
+  dif_hmac_endianness_t digest_endianness;
+  /** The HMAC key. Only used in |kDifHmacModeHmac| mode. */
+  uint32_t key[8];
```
Is there a reason this can't be a pointer instead? It feels somehow gross to me to pass a key inline.

Also, I'm somewhat worried about this being an array of words, rather than an array of bytes. This puts an alignment requirement on keys callers are passing in, which may result in unnecessary copies.

jon-flatley replied:
Since the key is optional in SHA256 mode I was trying to avoid a garbage pointer making its way in here. I can make it a pointer if that's preferable.

silvestrst commented:
I think you also need to run clang-format to silence the linter.

jon-flatley commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,221 @@
+  return kDifHmacFifoNoError;
+}
+
+dif_hmac_error_t dif_hmac_fifo_count_entries(const dif_hmac_t *hmac,
```
I think they're mostly equivalent, but this way has the added benefit that the caller can easily tell when the FIFO is empty before calling `dif_hmac_get_msg_len`. Maybe this could be made more clear by changing the function to `dif_hmac_fifo_empty` and returning just a `bool` out-parameter. I don't think there's any value in the caller knowing the exact number of FIFO entries.

jon-flatley commented on sw/device/tests/hmac/dif_hmac_test.c:
```diff
@@ -0,0 +1,189 @@
+    0xe4987b39, 0x3f83d390, 0xc2f3bbaf, 0x3195dbfa,
+    0x23fb480c, 0xb012ae5e, 0xf1394d28, 0x1940ceeb};
+
+static void print_uint(uint64_t n) {
```
`uart_send_uint` sends hex, which is fine for some things, but a bit cumbersome when you're just looking for lengths of data sent or an array index. `uart_send_uint` also prints a fixed length string by `size`, so it would yield error strings in the form 

> Digest mismatch, expected: 0x01234567 at index [00000001] but got: 0x89ABCDEF.

> Message length mismatch. Expected 000000AB bits but got 000000CD bits.

I just think it's cleaner to have decimal outputs with no leading zeros. Happy to move this to uart.[ch] if it's helpful.

silvestrst replied:
Oh, actually maybe it's worth using @mcy recently landed `print_log`. It is capable of printing ints and hex, not sure about 64bit though.

jon-flatley commented:
Thanks again for the comments. Changes since last push

- Split into two commits for the actual DIF and the DIF tests
- Refactored test into smaller functions with comments
- Used @mcy's log.h for string formatting
- Refactored various names
- Ran `clang-format`

alphan commented:
Yes, you’re right. For some reason I only thought about the beginning not
the end. Sorry for the confusion!

On Thu, Feb 13, 2020 at 1:30 PM jon-flatley <notifications@github.com>
wrote:

> *@jon-flatley* commented on this pull request.
> ------------------------------
>
> In sw/device/lib/dif/dif_hmac.c
> <https://github.com/lowRISC/opentitan/pull/1320#discussion_r379041524>:
>
> > +  return kDifHmacNoError;
> +}
> +
> +dif_hmac_fifo_error_t dif_hmac_fifo_push(const dif_hmac_t *hmac,
> +                                         const void *data, size_t len,
> +                                         size_t *bytes_sent) {
> +  if (hmac == NULL || data == NULL) {
> +    return kDifHmacFifoInvalidArgument;
> +  }
> +
> +  const uint8_t *data_sent = (uint8_t *)data;
> +  int words_remaining = len / sizeof(uint32_t);
> +  int bytes_remaining = len % sizeof(uint32_t);
> +
> +  // Write whole words until the FIFO fills or there are no words left to send.
> +  while (words_remaining > 0 && get_fifo_available_space(hmac) > 0) {
>
> I don't think so. Let's say len is 7 bytes, and aligned as such
>
> XXXXXX01 23456789 ABCDXXXX
>       ^
>   data_sent
>
> Since data_sent ins't necessarily aligned to begin with I think it would
> require three loops to
>
>    1. Bring data_sent into alignment by writing bytes from the first,
>    unaligned word { 0x01 }
>
> XXXXXX01 23456789 ABCDXXXX
>          ^
>      data_sent
>
>
>    1. Loop on aligned words with read_32 to send all 32-bit words {
>    0x23456789 }
>
> XXXXXX01 23456789 ABCDXXXX
>                   ^
>               data_sent
>
>
>    1. Write the final bytes from the last word { 0xAB, 0xCD }
>
> @mcy <https://github.com/mcy> is this what you're suggesting?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/lowRISC/opentitan/pull/1320?email_source=notifications&email_token=AN2D23SS4JRPQJJ5I3VTGPDRCWGUNA5CNFSM4KFOH3UKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCVOY2HQ#discussion_r379041524>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AN2D23XZQO4LD3ACM5BJXJLRCWGUNANCNFSM4KFOH3UA>
> .
>
-- 

Best,

alphan
_

jon-flatley commented:
Another push to address @mcy's [suggestion for word-aligned copies ](https://github.com/lowRISC/opentitan/pull/1320#discussion_r376600022) which now depends on #1547 to fix the read32() signature. Not sure what the best practices are for managing PR dependencies since GH doesn't seem to support them so please let me know if what I've done is reasonable.

jon-flatley commented:
Thanks @lenary for the comments.

Changes in latest push:
- Moved `HMAC_CMD_HASH_START` to a dedicated function, `dif_hmac_start()`
- Created a new error case for calls to `dif_hmac_fifo_push()` when `dif_hmac_start()` hasn't yet been called.
- Expanded tests cases to run multiple operations under a common configuration without re-initializing the `hmac` object (to test `dif_hmac_start()`.
- Resolved some nits

mcy commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,246 @@
+  }
+
+  // Check if hmac_done is asserted.
+  if (mmio_region_get_bit32(hmac->base_addr, HMAC_INTR_STATE_REG_OFFSET,
```
This is basically code golf at this point but how do you feel about
```c
*done = get_bit32();
if(*done) {
  set_bit32();
}
```
Something about gift-wrapping a `= bool` feels gross to me.

gkelly replied:
Something about declaring a new variable to only use it once feels gross to me. :wink:

mcy replied:
It's an out-param too, which is why its getting assigned in the first place.

jon-flatley commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,235 @@
+ */
+typedef struct dif_hmac {
+  mmio_region_t base_addr;
+  bool hmac_started;
```
Hey @lenary, thanks for your detailed comments. I've attempted to respond to your comments bellow. This really gets into some fundamental question about DIF design in general, so if others have thoughts on these items, I would love to hear them.

> After some reflection, I'm not sure we've found exactly the right abstraction level with this DIF yet. I have tried to provide concrete feedback below, but the overarching concern is that this DIF is operating at a level that's too high. I would like to see the abstraction level brought back down to the level of exposing hardware functionality without too much interference.
> 
> At some point we will be writing a higher-level driver which encapsulate some of the higher-level interfaces we aim for software to use when using the hardware module, and this is where things like state machine management should happen, not in the DIF.

I've also been struggling with finding the right abstraction level for this DIF. Is `dif_hmac_fifo_push()` too high level, for example? Maybe the looping logic should live in the higher level driver and the DIF should just expose 32-bit and 8-bit writes. A function to copy a buffer into a FIFO doesn't feel too high level, but maybe it is.

It's difficult for me to reason about where the DIF abstraction should stop. If the goal is to prevent the consumer of the DIF from misusing the hardware then the DIFs need to do some amount of heavy lifting. Just exposing hardware functionality doesn't address this. This is the most difficult property of DIFs for me to understand since DIFs need to be sophisticated enough to prevent misuse of the hardware, but simple enough that they aren't replacing high level driver functionality.

> This field is approximating a state machine, but is not fully representing all the possible states, given the state machine has to cope with the loading or trashing of the secret key, and the start/complete/retrieved digest states of a single hash operation.
> 
> There are two solutions here: remove the field, or track the state machine fully. I would suggest we remove the field, as tracking the full state machine seems like a driver issue, not a DIF issue.

I'm specifically tracking this bit because `dif_hmac_fifo_push()` can lock up if HMAC isn't running. Other functions, such as `dif_hmac_digest_read()` can produce incorrect results if called at the wrong time, but something about a DIF locking up feels like a problem in a totally different category. This is sort of in line with the philosophy behind argument checking; we check some arguments for reasonable values (e.g. nullness) but not others (e.g. enum ranges). I see checking for valid state in a similar light. Both checks are catching common cases where a function's contract has been violated. We've made the decision to check for certain kinds of invalid arguments but not others, so I thought it may also be reasonable to check for certain kinds of invalid state, especially when it can avoid deadlocks.

> I had a chat with @gkelly and, upon reflection, we're not sure having more than one dif_hmac_t for the same MMIO region is particularly safe (it opens up too many concurrency issues).
> 
> This almost certainly suggests, as @silvestrst says, that there should be a way, in this DIF, to:
> 
> choose a different HMAC mode, endianness configuration,
> enable disable both the HMAC datapath and the SHA datapath
> load in a specific secret key
> These should be different from initialisation, and preferably different from each other.

I agree that multiple concurrent dif_hmac_t objects is recipe for disaster.

The `dif_hmac_init()` function is a bit of a catch-all. I could break these things out into separate functions, but this would probably be a replacement for `dif_hmac_config`, which already has all of these options.

Is there a reason to support disabling the SHA datapath? HMAC mode enables both.

mcy replied:
My personal theory on the abstraction boundary is thus.

Deivce X is basically a FIFO with some kind of property.
- The DIF *should* provide a "try to copy as much as possible to this FIFO" function.
- The DIF *should not* provide multiplexing and scheduling for that FIFO: that's the operating system's job.

I think that the DIF should maintain whatever state the programmer's guide for the device spec states, and no more. The view the DIF provides is as a single, global FIFO, which the operating system can safely consume without knowing the underlying details, other than that it is a FIFO that does things to the data. The view the operating system provides is the "FIFO as a service": all of the processes get to pretend they have their own FIFO, perhaps with some kind of scheduling or eviction constraints that make it possible for the operating system to multiplex the resource.

jon-flatley commented:
New push to address @lenary's comments:
- Removed excessive state tracking inside DIF. This should no longer be needed now that #1869 has been resolved.
- Added new functions to change the datapath (HMAC vs SHA256 only), message endianness, digest endianness, and secret key without having to reinitialize the entire DIF object.
- Updated error handing slightly to include a `kDifHmacErrorUnknown` error. This is only used for invalid `default` cases on enum switches.

Hopefully this gets us closer to the correct level of abstraction here.

alphan commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,339 @@
+/**
+ * Load |key| into |hmac|.
+ */
+static void load_key(const dif_hmac_t *hmac, const uint8_t *key) {
+  for (int i = 0; i < HMAC_KEY_REG_COUNT; ++i) {
+    ptrdiff_t offset = i * sizeof(uint32_t);
+    uint32_t key_part;
+
+    memcpy(&key_part, key + offset, sizeof(key_part));
+
+    mmio_region_write32(hmac->base_addr, HMAC_KEY0_REG_OFFSET + offset,
+                        key_part);
+  }
+}
+
```
Not related to this PR, but I was wondering how you would feel about replacing this with `mmio_region_memcpy_to_mmio32` once #2066 lands.

jon-flatley replied:
#2066 will make this much cleaner, I'll take note to refactor this once it gets merged, thanks!

alphan commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,339 @@
+  }
+
+  // Read the digest.
+  for (int i = 0; i < HMAC_DIGEST_REG_COUNT; ++i) {
+    ptrdiff_t offset = i * sizeof(uint32_t);
+    digest->digest[i] =
+        mmio_region_read32(hmac->base_addr, HMAC_DIGEST0_REG_OFFSET + offset);
+  }
```
This too.

jon-flatley commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,289 @@
+ * referenced by |digest_out|. |dif_hmac_state()| should before this
+ * function to ensure the digest is available to be read.
+ *
+ * |digest_out| must reference an allocated, contiguous, 32-byte buffer.
```
I am by no means an expert on alignment. Am I correct in my belief that `dif_hmac_digest_t` must be 4-byte aligned since `uint_32` is its largest member, and that this is the only alignment requirement?

lenary replied:
You're approximately correct.

In C, arrays are aligned to the same alignment as their element type. C does not specify the alignment of structs, but various ABIs do, and in short structs get aligned to the alignment of their _first_ member (not their largest, I get caught out by this all the time). All scalars in C are aligned to at least their size (this is called "natural alignment").

jon-flatley commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,289 @@
+ * the FIFO fills up or |len| bytes have been sent.
+ *
+ * In the event that the FIFO fills up before |len| bytes have been sent this
+ * function will return a |kDifHmacFifoErroFull| error.
```
Done

jon-flatley commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,289 @@
+ *
+ * Calling this function has the side effect of clearing the hmac_done status,
+ * meaning that after this function returns true subsequent calls will return
+ * false until |dif_hmac_run()| is called again.
```
Done.

jon-flatley commented on sw/device/lib/dif/meson.build:
```diff
@@ -55,3 +55,16 @@ sw_lib_dif_spi_device = declare_dependency(
+      hw_ip_hmac_reg_h,
+      'dif_hmac.c'
+    ],
+    dependencies: [sw_lib_mem],
```
Done.

jon-flatley commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,339 @@
+#include "sw/device/lib/dif/dif_hmac.h"
+
+#include "hmac_regs.h"  // Generated.
+#include "sw/device/lib/base/memory.h"
```
Done.

jon-flatley commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,339 @@
+
+#define HMAC_HASH_SIZE_BITS 256u
+#define HMAC_KEY_REG_COUNT (HMAC_HASH_SIZE_BITS / 32)
+#define HMAC_DIGEST_REG_COUNT HMAC_KEY_REG_COUNT
```
Done.

jon-flatley commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,339 @@
+  const uint8_t *data_sent = (const uint8_t *)data;
+
+  int unaligned_bytes = (uintptr_t)data_sent % sizeof(uint32_t);
+  int words_remaining = (len - unaligned_bytes) / sizeof(uint32_t);
+  int bytes_remaining = (len - unaligned_bytes) % sizeof(uint32_t);
```
Fixed.

jon-flatley commented:
New push to address @lenary's comments:
- refactored onto master, necessitating moving the smoke test to accommodate the new structure of `sw/device/tests`
- now uses the error format described in #2212 
- fixes bug where a buffer overrun was possible on messages of length < 4 passed to `dif_hmac_fifo_push()`
- resolved outdated documentation and other nits

This particular push is untested as of right now, I will likely get around to that tomorrow once my FPGA bitstream finishes building. It also looks like `clang-format` isn't happy so there will be another small push after this to clean this up.

lenary commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,343 @@
+  }
+
+  if (bytes_sent != NULL) {
+    *bytes_sent = len - (words_remaining * sizeof(uint32_t)) - bytes_remaining;
```
Is this calculation correct? what if `get_fifo_available_space(hmac) == 0` when `unaligned_bytes` is not yet zero?

jon-flatley replied:
Another great catch, thanks!

jon-flatley commented:
New push, the implementation of `dif_hmac_fifo_push()` was problematic, and fixing it was quickly adding complexity. I've refactored it to look more like @alphan's suggestion with a single loop.

mcy commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,329 @@
+#include "sw/device/lib/base/memory.h"
+#include "sw/device/lib/base/mmio.h"
+
+#define HMAC_HASH_SIZE_BITS 256u
```
Unused?

mcy commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,329 @@
+#define HMAC_HASH_SIZE_BITS 256u
+
+/**
+ * Load |key| into |hmac|.
```
Might want to update `|code|` to `` `code` `` throughout.

mcy commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,329 @@
+/**
+ * Load |key| into |hmac|.
+ */
+static void load_key(const dif_hmac_t *hmac, const uint8_t *key) {
+  for (int i = 0; i < HMAC_PARAM_NUMWORDS; ++i) {
+    ptrdiff_t offset = i * sizeof(uint32_t);
+    uint32_t key_part;
+
+    memcpy(&key_part, key + offset, sizeof(key_part));
+
+    mmio_region_write32(hmac->base_addr, HMAC_KEY0_REG_OFFSET + offset,
+                        key_part);
+  }
+}
```
Maybe you should consider using the new MMIO `memcpy` for this?

mcy commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,329 @@
+ * @return The number of entries in the HMAC FIFO.
+ */
+static uint32_t get_fifo_entry_count(const dif_hmac_t *hmac) {
+  return (get_status(hmac) >> HMAC_STATUS_FIFO_DEPTH_OFFSET) &
+         HMAC_STATUS_FIFO_DEPTH_MASK;
```
This might read better as a `bitfield_get_field32` call.

mcy commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,329 @@
+    return kDifHmacFifoBadArg;
+  }
+
+  const uint8_t *data_sent = (const uint8_t *)data;
+  int bytes_remaining = len;
+
+  while (bytes_remaining > 0 && get_fifo_available_space(hmac) > 0) {
+    bool word_aligned = (uintptr_t)data_sent % sizeof(uint32_t) == 0;
+    int bytes_written = 0;
+
+    if (bytes_remaining < sizeof(uint32_t) || !word_aligned) {
+      // Individual byte writes are needed if the buffer isn't aligned or there
+      // are no more full words to write.
+      mmio_region_write8(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, *data_sent);
+      bytes_written = 1;
+    } else {
+      // |data_sent| is word-aligned and there are still words to write.
+      uint32_t word = read_32(data_sent);
+      mmio_region_write32(hmac->base_addr, HMAC_MSG_FIFO_REG_OFFSET, word);
+      bytes_written = sizeof(uint32_t);
+    }
+
+    bytes_remaining -= bytes_written;
+    data_sent += bytes_written;
+  }
```
This also looks like it wants `mmio_region_memcpy()`.

lenary replied:
I think doing so will be incorrect here.

`region_memcpy` will do a read modify write on the last write, ensuring the last write is 32 bits wide, regardless of the width of the original buffer (so the length of the total written buffer becomes a multiple of 4 bytes). However, in HMAC, the packer logic is calculating the length based on the writes it is seeing, so doing a single byte-write is actually a different operation to doing a word-write (one increases the hashed length by 1, the other by 4.

alphan replied:
[I think so too](https://github.com/lowRISC/opentitan/pull/2066#issuecomment-622673045).

mcy commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,294 @@
+ * Configuration for initializing the HMAC device.
+ */
+typedef struct dif_hmac_config {
+  /** The base address for registers in the HMAC IP*/
+  mmio_region_t base_addr;
+  /** The mode of operation */
+  dif_hmac_mode_t mode;
+  /** Byte endianness for writes to the FIFO */
+  dif_hmac_endianness_t message_endianness;
+  /** Byte endianness for reads from the digest */
+  dif_hmac_endianness_t digest_endianness;
+  /** The HMAC key. Only used in |kDifHmacModeHmac| mode. Must point to 256 bits
+   * of contiguous, allocated memory */
```
Missing periods on these comments.

mcy commented on sw/device/tests/dif/dif_hmac_device_test.c:
```diff
@@ -0,0 +1,290 @@
+      case kDifHmacOk:
+        break;
+      case kDifHmacBadArg:
+        LOG_ERROR("Invalid arguments encountered in HMAC init.\r\n");
```
You can drop `\r\n` in `LOG` calls.

mcy commented on sw/device/tests/dif/dif_hmac_device_test.c:
```diff
@@ -0,0 +1,290 @@
+static bool run_test_impl(const dif_hmac_config_t *config, const char *data,
+                          size_t len, const uint32_t *expected_digest) {
+  dif_hmac_t hmac;
+  if (!test_setup(config, &hmac)) {
```
These can all be replaced with `CHECK(test_setup(...))`, etc.

mcy commented on sw/device/tests/dif/dif_hmac_device_test.c:
```diff
@@ -0,0 +1,290 @@
+  }
+}
+
+int main(int argc, char **argv) {
```
I believe Sri's `test_main` PR landed, so I would consider refactoring this to use that, instead.

gkelly commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,329 @@
+  // Set the byte-order of the input message.
+  switch (config->message_endianness) {
+    case kDifHmacEndiannessBig:
+      device_config |= (1 << HMAC_CFG_ENDIAN_SWAP);
```
This depends on host endianness, so it's not 100% accurate to say that we have to swap endianness to be big-endian. That being said, without doing gross pointer stuff or depending on #defines to give us this information I don't know that we can be fully accurate here. Can you add a TODO in here to address that? We can commit with the TODO I just want to make sure we fix it up later.

gkelly commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,329 @@
+  return kDifHmacOk;
+}
+
+dif_hmac_result_t dif_hmac_done(const dif_hmac_t *hmac, bool *done) {
+  if (hmac == NULL || done == NULL) {
+    return kDifHmacBadArg;
+  }
+
+  // Check if hmac_done is asserted.
+  *done = mmio_region_get_bit32(hmac->base_addr, HMAC_INTR_STATE_REG_OFFSET,
+                                HMAC_INTR_ENABLE_HMAC_DONE);
+  if (*done) {
+    // Clear hmac_done.
+    mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_INTR_STATE_REG_OFFSET,
+                                    HMAC_INTR_ENABLE_HMAC_DONE);
+  }
+
+  return kDifHmacOk;
+}
+
+dif_hmac_result_t dif_hmac_digest_read(const dif_hmac_t *hmac,
```
Can you merge these functions into a single one that checks for done and populates `digest` if it's complete? The use pattern here seems to be just always calling `_done` and then calling `_digest_read`, so merging them seems sensible.

gkelly commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,294 @@
+ * @return |kHmacErrorInvalidArgument| if |hmac| is null |kHmacErrorNone|
+ * otherwise.
+ */
+dif_hmac_result_t dif_hmac_run(dif_hmac_t *hmac);
```
`dif_hmac_run` -> `dif_hmac_process`, to match the hardware naming.

jon-flatley commented:
Thanks @gkelly, @mcy, and @lenary for the most recent round of detailed reviews. I believe I have addressed everything.

- Added interrupt API.
- Removed `key` and `mode` from the config.
- Removed individual config functions.
- Added mode specific `_start()` functions.
- Refactored tests to use `test_main()`.
- Updated documentation style.

silvestrst commented on sw/device/tests/meson.build:
```diff
@@ -67,6 +78,7 @@ sw_tests += {
   'aes_test': aes_test_lib,
   'flash_ctrl_test': flash_ctrl_test_lib,
   'rv_timer_test': rv_timer_test_lib,
+  'hmac_dif_test': hmac_dif_test_lib,
```
should be `dif_hmac_sanitytest` and live in `.../tests/dif`

silvestrst commented on sw/device/tests/hmac_dif_test.c:
```diff
@@ -148,143 +159,95 @@ static bool check_message_length(const dif_hmac_t *hmac,
-  } else {
-    LOG_INFO("FAIL!\r\n");
-  }
+  CHECK(test_setup(config, &hmac, key));
```
If you use `CHECK` inside the lower level functions, you won't need to check it here. `CHECK` does `abort` internally, it would then simplify things slightly.

silvestrst commented on sw/device/tests/hmac_dif_test.c:
```diff
@@ -148,143 +159,95 @@ static bool check_message_length(const dif_hmac_t *hmac,
-          "Digest mismatch, expected: 0x%X at index [%i] but got: "
-          "0x%X.\r\n",
-          expected_digest[i], i, digest_result.digest[i]);
+      LOG_ERROR("Digest mismatch, expected: 0x%X at index [%i] but got: 0x%X.",
```
I guess this could be directly turned into `CHECK`, and this function can become `void`. The same probably applies to other similar functions.

Please see the comment in `run_test`.

jon-flatley commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -99,110 +120,148 @@ dif_hmac_result_t dif_hmac_init(const dif_hmac_config_t *config,
+  if (hmac == NULL) {
+    return kDifHmacBadArg;
+  }
+
```
Turns out there are a couple of issues with these new start functions, I believe I've addressed everything now.

lenary replied:
I think it would be better to have a `dif_hmac_finish` which combines `dif_hmac_digest_read` and clearing `CFG.SHA_EN` and `CFG.HMAC_EN`. Then it becomes clearer when a transaction is in progress (between the start and finish calls). It would mean you only had one opportunity to read the digest, but I don't think that's a problem, if you need it for longer read it into a buffer and pass that around.

Let's file this in the follow-up issue, not to block this PR.

jon-flatley commented:
Thanks for the reviews folks, and thanks for sticking through this long series of changes.

I've most recently addressed:
- Test location, now located under `sw/device/tests/dif/dif_hmac_sanitytest.c`
- Tests have been refactored to use `CHECK`s closer to the points of failure rather than towards the top level
- Fixed some issues with the `_start()` functions, including clearing the `HMAC_CFG_HMAC_EN` bit
- Squashed back into two commits

alphan commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,364 @@
+/**
+ * State for a particular HMAC device.
+ *
+ * its member variables should be considered private, and are only provided so
```
Capialization: its

alphan commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,400 @@
+                               ptrdiff_t *offset_out) {
+  ptrdiff_t offset;
+  switch (irq_type) {
+    case kDifHmacInterruptHmacDone:
+      offset = HMAC_INTR_STATE_HMAC_DONE;
+      break;
+    case kDifHmacInterruptFifoEmpty:
+      offset = HMAC_INTR_STATE_FIFO_EMPTY;
+      break;
+    case kDifHmacInterruptHmacErr:
+      offset = HMAC_INTR_STATE_HMAC_ERR;
```
You may want to use `HMAC_INTR_COMMON_*` for the sake of clarity.

alphan commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,400 @@
+
+  // Clear INTER_STATE.
+  mmio_region_write32(config->base_addr, HMAC_INTR_STATE_REG_OFFSET,
+                      (1 << HMAC_INTR_ENABLE_HMAC_DONE) |
```
Same value, but should be `HMAC_INTR_STATE_HMAC_DONE` for the sake of consistency.

alphan commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,400 @@
+  // Clear the config, stopping the SHA engine.
+  mmio_region_write32(config->base_addr, HMAC_CFG_REG_OFFSET, 0);
+
+  // Clear INTER_STATE.
```
typo: `INTER`

alphan commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,400 @@
+    return kDifHmacBadArg;
+  }
+
+  // Pass the interrupt state back to the caller.
+  if (state != NULL) {
+    *state = mmio_region_read32(hmac->base_addr, HMAC_INTR_STATE_REG_OFFSET);
+  }
+
+  // Disable all interrupts.
+  mmio_region_write32(hmac->base_addr, HMAC_INTR_STATE_REG_OFFSET, 0u);
```
Wrong register, should access `HMAC_INTR_ENABLE_REG_OFFSET`.

alphan commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,400 @@
+  }
+
+  // Restore interrupt state.
+  mmio_region_write32(hmac->base_addr, HMAC_INTR_STATE_REG_OFFSET, state);
```
Wrong register, should write to `HMAC_INTR_ENABLE_REG_OFFSET`.

alphan commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,400 @@
+    return kDifHmacError;
+  }
+
+  // Enable/Disable interrupt state.
+  if (enable == kDifHmacEnable) {
+    mmio_region_nonatomic_set_bit32(hmac->base_addr, HMAC_INTR_STATE_REG_OFFSET,
+                                    offset);
+  } else {
+    mmio_region_nonatomic_clear_bit32(hmac->base_addr,
+                                      HMAC_INTR_STATE_REG_OFFSET, offset);
+  }
```
Wrong register, should write to `HMAC_INTR_ENABLE_REG_OFFSET`. Also, please consider dropping "state" from the comment.

alphan commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,400 @@
+ * Convert from a `dif_hmac_interrupt_t` to the appropriate bit index.
+ * INTR_STATE, INTR_ENABLE, and INTR_TEST registers have the same bit offset.
+ */
+static bool irq_bit_offset_get(dif_hmac_interrupt_t irq_type,
+                               ptrdiff_t *offset_out) {
```
I'm not sure `ptrdiff_t` is the best type to use here since these are bit offsets. `uint8_t` could be better.

alphan commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,400 @@
+  }
+
+  const uint8_t *data_sent = (const uint8_t *)data;
+  int bytes_remaining = len;
+
+  while (bytes_remaining > 0 && get_fifo_available_space(hmac) > 0) {
+    bool word_aligned = (uintptr_t)data_sent % sizeof(uint32_t) == 0;
+    int bytes_written = 0;
```
I think, since `len` and `bytes_sent` are `size_t`, `bytes_remaining` and `bytes_written` should also be `size_t`.

alphan commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,400 @@
+ * Load `key` into `hmac`.
+ */
+static void load_key(const dif_hmac_t *hmac, const uint8_t *key) {
+  mmio_region_memcpy_to_mmio32(hmac->base_addr, HMAC_KEY0_REG_OFFSET, key,
+                               HMAC_PARAM_NUMWORDS * sizeof(uint32_t));
```
Since this requires the addresses of `HMAC_KEY0_REG_OFFSET` ... `HMAC_KEY7_REG_OFFSET` to be consecutive, you may want to add an assert here or in the unit test.

lenary replied:
`_Static_assert` is probably the way forwards, I'm happy for that to happen in a follow-up.

alphan commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,400 @@
+  mmio_region_nonatomic_clear_bit32(hmac->base_addr, HMAC_CFG_REG_OFFSET,
+                                    HMAC_CFG_SHA_EN);
+
+  load_key(hmac, key);
```
Can be inlined since `load_key` is not used anywhere else.

alphan commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,400 @@
+  }
+
+  // Read the digest.
+  for (int i = 0; i < HMAC_PARAM_NUMWORDS; ++i) {
+    ptrdiff_t offset = i * sizeof(uint32_t);
+    digest->digest[i] =
+        mmio_region_read32(hmac->base_addr, HMAC_DIGEST0_REG_OFFSET + offset);
+  }
```
You may want to use `mmio_region_memcpy_from_mmio32`, similar to `load_key` but in the other direction.

lenary commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,364 @@
+ * https://docs.opentitan.org/doc/rm/comportability_specification/
+ */
+typedef enum dif_hmac_interrupt {
+  //! HMAC is done.
```
Can you file a follow-up issue to go through and replace your `//!` and `/*!` comments with `/**` comments? Let's not block this PR, though.

gkelly commented:
> I've just commented about a few things that I think we should change, but should do in a follow-up. Can you be sure to file an issue with a todo list covering everything currently outstanding?
> 
> Another thing we should cover in the follow-up is moving to the new bitfield functions.

Agreed, these are good candidates for a followup PR.

mcy commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,400 @@
+ * Convert from a `dif_hmac_interrupt_t` to the appropriate bit index.
+ * INTR_STATE, INTR_ENABLE, and INTR_TEST registers have the same bit offset.
+ */
+static bool irq_bit_offset_get(dif_hmac_interrupt_t irq_type,
```
This function (and other error-like returning functions) probably wants to be marked as DIF_WARN_UNUSED_RESULT. You should do this in a followup. (Also, might want to make this return `kBadArg` instead of `false`...?)

mcy commented on sw/device/lib/dif/dif_hmac.c:
```diff
@@ -0,0 +1,400 @@
+  mmio_region_write32(config->base_addr, HMAC_CFG_REG_OFFSET, 0);
+
+  // Clear INTER_STATE.
+  mmio_region_write32(config->base_addr, HMAC_INTR_STATE_REG_OFFSET,
+                      (1 << HMAC_INTR_ENABLE_HMAC_DONE) |
+                          (1 << HMAC_INTR_STATE_FIFO_EMPTY) |
+                          (1 << HMAC_INTR_STATE_HMAC_ERR));
```
In a followup, consider using bitfield.h to build up this value instead. (This applies to all other bit manips in this file).

mcy commented on sw/device/lib/dif/dif_hmac.h:
```diff
@@ -0,0 +1,364 @@
+ * @return `kDifHmacBadArg` if `config` is null or contains illegal
+ *          arguments or `hmac_out` is null, `kDifHmacOk` otherwise.
+ */
+dif_hmac_result_t dif_hmac_init(const dif_hmac_config_t *config,
```
In a followup, please mark all of these as DIF_WARN_UNUSED_RESULT

mcy commented on sw/device/tests/dif/dif_hmac_sanitytest.c:
```diff
@@ -0,0 +1,203 @@
+static void test_setup(const dif_hmac_config_t *config, dif_hmac_t *hmac) {
+  dif_hmac_result_t res = dif_hmac_init(config, hmac);
+
+  CHECK(res != kDifHmacBadArg, "Invalid arguments encountered in HMAC init.");
+  CHECK(res == kDifHmacOk, "Unknown error encountered in HMAC init.");
```
I don't think you should be duplicating these CHECK like this. You can clean this up in a followup.

mcy commented on sw/device/tests/dif/dif_hmac_sanitytest.c:
```diff
@@ -0,0 +1,203 @@
+                         const uint32_t *expected_digest) {
+  dif_hmac_digest_t digest_result;
+  bool hmac_done = false;
+  do {
+    dif_hmac_digest_result_t res = dif_hmac_digest_read(hmac, &digest_result);
+
+    CHECK(res != kDifHmacDigestBadArg,
+          "Invalid arguments encountered reading HMAC digest.");
+
+    hmac_done = (res != kDifHmacDigestProcessing);
+
+    if (hmac_done) {
+      CHECK(res == kDifHmacDigestOk,
+            "Unknown error encountered reading HMAC digest.");
+    }
+  } while (!hmac_done);
```
AFAICT this can just be written as a while() loop, which is less confusing than a do/while. You can punt this to a followup.

mcy commented on sw/device/tests/dif/dif_hmac_sanitytest.c:
```diff
@@ -0,0 +1,203 @@
+    CHECK(res != kDifHmacDigestBadArg,
+          "Invalid arguments encountered reading HMAC digest.");
+
+    hmac_done = (res != kDifHmacDigestProcessing);
```
These parentheses are redundant.

jon-flatley replied:
They're nice to have IMO for the sake of clarity.

jon-flatley commented:
I created an issue to track outstanding issues mentioned in this PR, see #2920.

Aside from those items all comments have been addressed and resolved.

jon-flatley commented:
Looks like an unstaged file from another branch ended up here, removing that and fixing a header license issue.

lenary commented:
:tada:
